### PR TITLE
MES-2547 - Rekey

### PR DESCRIPTION
**Pull request by roberthall12**
## Description and relevant Jira numbers

Created a rekey action that is called in the `ActivateTest` and `StartTest` effects that sets the `rekey` field in the test. This occurs either when the test is before 00:00 today or the time is after the `start time + duration of test` a modal is displayed to select if rekey or not.
https://jira.i-env.net/browse/MES-2547

## Pull Request checklist:

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist:

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop
- [x] commits are squashed

## Sign off process checklist:

- [x] Code has been tested manually
- [x] Tests are passing
- [x] PR link added to JIRA
- [ ] Reviewers selected in Github
- [ ] Any new reusable component/widget added to component library on Confluence

- [ ] Screenshot(s) captured on the physical device (if applicable)
- [ ] Tested by QA
- [ ] PO's approval

Screenshot 2019-07-02 at 16 35 00
Screenshot 2019-07-02 at 16 34 45

### CODE REVIEW

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 1]
```diff
@@ -76,23 +71,83 @@ export class TestOutcomeComponent {
...
+      this.testStatus === TestStatus.Submitted
+    ) {
+      return false;
+    }
+    const testDate = new Date(this.slotDetail.start).setUTCHours(0, 0, 0, 0);
+    const todaysDate = new Date().setUTCHours(0, 0, 0, 0);
+    if (testDate < todaysDate) {
+      return true;
+    }
+    return false;
```
Comment (buzzub) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: Could shorten this to `return testDate < todaysDate`

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 2]
```diff
@@ -76,23 +71,83 @@ export class TestOutcomeComponent {
...
+  }
+
+  showRekeyModal(): boolean {
+    const cutOffTime = new Date(this.slotDetail.start);
+    cutOffTime.setMinutes(cutOffTime.getMinutes() + this.slotDetail.duration);
+    const todaysDateTime = new Date();
+    if (todaysDateTime > cutOffTime) {
+      return true;
+    }
+    return false;
```
Comment (buzzub) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: Could shorten this to `return todaysDateTime > cutOffTime`

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 3]
```diff
@@ -76,23 +71,83 @@ export class TestOutcomeComponent {
...
+  }
+
   onModalDismiss = (event: ModalEvent): void => {
     switch (event) {
       case ModalEvent.CANCEL:
         this.modal.dismiss();
         break;
       case ModalEvent.START:
-        // todo: Logic to be impliment later
+        console.log('start');
```
Comment (rstepnowski) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: That console.log can be removed now, right?

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 4]
```diff
@@ -76,23 +71,83 @@ export class TestOutcomeComponent {
...
+        this.startTest(false);
         break;
       case ModalEvent.REKEY:
-        // todo: Logic to be impliment later
+        console.log('rekey');
```
Comment (rstepnowski) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: same here

**Comment by ammarhaiderbjss** [comment]
@roberthall12 quick screenshot?

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 5]
```diff
@@ -17,7 +19,7 @@ import { ModalEvent } from '../../journal-rekey-modal/journal-rekey-modal.consta
 export class TestOutcomeComponent {
 
   @Input()
-  slotId: number;
+  slotDetail: any;
```
Comment (ammarhaiderbjss) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: Now that the journal schema is centralised, can we improve this typing?

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 6]
```diff
@@ -47,22 +49,18 @@ export class TestOutcomeComponent {
...
-
   writeUpTest() {
-    this.store$.dispatch(new ActivateTest(this.slotId));
+    this.store$.dispatch(new ActivateTest(this.slotDetail.slotId));
     this.navController.push(OFFICE_PAGE);
   }
 
   resumeTest() {
-    this.store$.dispatch(new ActivateTest(this.slotId));
+    if (this.showRekeyModal()) {
```
Comment (ammarhaiderbjss) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: the name of this method is a bit too similar to `displayRekeyModal`. They're a bit confusing when the appear together like this if statement. Can we rename so they're distinct?

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 7]
```diff
@@ -74,23 +72,74 @@ export class TestOutcomeComponent {
...
       case ModalEvent.REKEY:
-        // todo: Logic to be impliment later
+        this.startTest(true);
         break;
     }
   }
+
+  rekeyTest() {
+    if (this.testStatus === TestStatus.Booked) {
+      this.startTest(true);
```
Comment (ammarhaiderbjss) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: How come the test is a rekey if it was booked? Shouldn't this use the logic from the `showRekeyModal()` method?

Comment (roberthall12) [reply src/pages/journal/components/test-outcome/test-outcome.ts]: The `showRekeyModal()` logic is only for tests today that are before the `start time + duration of test`. E.g. if it's a 10am test and 60 min duration the modal should show from 11am onwards. The blue rekey button is displayed if the test is on a previous day.

File: src/pages/journal/components/test-outcome/test-outcome.ts [hunk 8]
```diff
@@ -86,11 +89,54 @@ export class TestOutcomeComponent {
...
         break;
     }
   }
+
+  shouldDisplayRekeyModal(): boolean {
+    return this.isTestIncomplete() && this.isTodaysDate() && this.hasTestTimeFinished();
+  }
+
+  clickStartOrResumeTest() {
+    console.log('function 4');
```
Comment (jameshighfield) [inline src/pages/journal/components/test-outcome/test-outcome.ts]: Lingering log here

Comment (roberthall12) [reply src/pages/journal/components/test-outcome/test-outcome.ts]: good catch

File: src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts [hunk 9]
```diff
@@ -167,5 +163,77 @@ describe('Test Outcome', () => {
       });
     });
 
+    describe('show rekey button', () => {
+      it('should show rekey button for a booked test if date is in past', () => {
+        component.slotDetail = testSlotDetail;
+
+        const dateTime = new Date();
+        dateTime.setDate(dateTime.getDate() - 1);
```
Comment (rstepnowski) [inline src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts]: Please use the same logic to manipulate dates as in the production code (our abstraction is used)
It might be better to fix the date instead of ```new Date()``` and cover the edge cases in separate scenarios

File: src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts [hunk 10]
```diff
@@ -167,5 +163,77 @@ describe('Test Outcome', () => {
...
+
+        const rekeyButton = fixture.debugElement.query(By.css('.mes-rekey-button'));
+
+        expect(rekeyButton).not.toBeNull();
+      });
+      it('should show rekey button for a resumed test if date is in past', () => {
+        component.slotDetail = testSlotDetail;
+
+        const dateTime = new Date();
+        dateTime.setDate(dateTime.getDate() - 1);
```
Comment (rstepnowski) [inline src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts]: same as above - date manipulation stuff

File: src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts [hunk 11]
```diff
@@ -167,5 +163,77 @@ describe('Test Outcome', () => {
...
+        const rekeyButton = fixture.debugElement.query(By.css('.mes-rekey-button'));
+
+        expect(rekeyButton).not.toBeNull();
+      });
+      it('should hide rekey button for a booked test if date is today', () => {
+        component.slotDetail = testSlotDetail;
+
+        const dateTime = new Date();
+
+        component.slotDetail.start = dateTime.toUTCString();
```
Comment (rstepnowski) [inline src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts]: UI should not rely on UTC time

File: src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts [hunk 12]
```diff
@@ -167,5 +163,77 @@ describe('Test Outcome', () => {
...
+
+        const rekeyButton = fixture.debugElement.query(By.css('.mes-rekey-button'));
+
+        expect(rekeyButton).toBeNull();
+      });
+      it('should display the rekey modal for a test today that has ended', () => {
+        component.slotDetail = testSlotDetail;
+
+        const dateTime = new Date();
+        dateTime.setHours(dateTime.getHours() - 2);
```
Comment (rstepnowski) [inline src/pages/journal/components/test-outcome/__tests__/test-outcome.spec.ts]: Similar comment  var.setDate(getDate() -1) -> it might go to negative values depending on the date/time when the test is executed

File: src/modules/tests/__tests__/tests.selector.spec.ts [hunk 13]
```diff
@@ -136,6 +137,7 @@ describe('testsSelector', () => {
       testState.activityCode = ActivityCodes.CANDIDATE_NOT_HAPPY_WITH_AUTHORISED_OCCUPANT;
       const result = getTestOutcomeText(testState);
       expect(result).toBe(TestOutcome.Terminated);
+      expect(result).toBe('Terminated');
```
Comment (Cavalier07) [inline src/modules/tests/__tests__/tests.selector.spec.ts]: Is this the same as the line above

File: src/modules/tests/rekey/__tests__/rekey.reducer.spec.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import { rekeyReducer } from '../rekey.reducer';
+import { MarkAsRekey } from '../rekey.actions';
+
+it('should return true if MarkAsRekey action passed', () => {
```
Comment (Cavalier07) [inline src/modules/tests/rekey/__tests__/rekey.reducer.spec.ts]: Test isn't run if it's not in a describe

File: src/modules/tests/rekey/rekey.actions.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+import { Action } from '@ngrx/store';
+
+export const MARK_AS_REKEY = '[Rekey Actions] Mark the test as being rekeyed';
+
+export class MarkAsRekey implements Action {
+  readonly type = MARK_AS_REKEY;
+}
+
+export type Types =
+  | MarkAsRekey;
```
Comment (Cavalier07) [inline src/modules/tests/rekey/rekey.actions.ts]: Can remove the |

File: src/modules/tests/tests.actions.ts [hunk 16]
```diff
@@ -17,6 +17,8 @@ export const LOAD_PERSISTED_TESTS_SUCCESS = '[Tests] Load persisted success';
 export const SET_ACTIVITY_CODE = '[Tests] Set activity code';
 export const START_TEST_REPORT_PRACTICE_TEST = '[Tests] Start practice test';
 
+export const MARK_AS_REKEY = '[Tests] Mark test as rekey';
```
Comment (Cavalier07) [inline src/modules/tests/tests.actions.ts]: Not used

File: src/pages/journal/journal.effects.ts [hunk 17]
```diff
@@ -9,7 +9,7 @@ import { groupBy, has } from 'lodash';
 
 import * as journalActions from './journal.actions';
```
Comment (Cavalier07) [inline src/pages/journal/journal.effects.ts]: Missing tests for the new effect

File: package.json [hunk 18]
```diff
@@ -108,7 +108,7 @@
     "@angular/compiler-cli": "^5.2.11",
     "@dvsa/mes-journal-schema": "^1.0.7",
     "@dvsa/mes-search-schema": "^1.0.2",
-    "@dvsa/mes-test-schema": "^1.0.21",
+    "@dvsa/mes-test-schema": "^1.0.22",
```
Comment (johnmorleybjss) [inline package.json]: Conflicts here, we're already on `1.0.23`.

File: src/pages/journal/journal.effects.ts [hunk 19]
```diff
@@ -261,4 +268,13 @@ export class JournalEffects {
       costCode: slotData.testCentre.costCode,
     };
   }
+
+  @Effect()
+  activateTestCompleteEffect$ = this.actions$.pipe(
```
Comment (johnmorleybjss) [inline src/pages/journal/journal.effects.ts]: Would `activateTestEffect` be a better name?
